File: LeetCode/ProblemSet/Volume (300-399)/LeetCode316/LeetCode316-Solution-02.py
from typing import *
from collections import Counter

class Solution:
    def removeDuplicateLetters(self, s: str) -> str:
        cnt = Counter(s)
        book = set()
        ans = []

        for c in s:
            cnt[c] -= 1

            if c in book:
                continue

            while ans and c < ans[-1] and cnt[ans[-1]] > 0:
                book.remove(ans[-1])
                ans.pop()

            book.add(c)
            ans.append(c)

        return ''.join(ans)

File: LeetCode/ProblemSet/Volume (300-399)/LeetCode316/LeetCode316-Solution-02.go
package leetcode316

func removeDuplicateLetters_02(s string) string {
	cnt := make(map[rune]int)
	for _, c := range s {
		cnt[c]++
	}

	book := make(map[rune]struct{})
	ans := make([]rune, 0)
	for _, c := range s {
		cnt[c]--
		if _, exist := book[c]; exist {
			continue
		}

		for len(ans) > 0 && c < ans[len(ans)-1] && cnt[ans[len(ans)-1]] > 0 {
			delete(book, ans[len(ans)-1])
			ans = ans[:len(ans)-1]
		}

		ans = append(ans, c)
		book[c] = struct{}{}
	}

	return string(ans)
}
